pkg/autoscaler/collector.go

```diff
@@ -271,24 +291,36 @@ func (c *collection) currentMetric() *av1alpha1.Metric {
 // record adds a stat to the current collection.
 func (c *collection) record(stat Stat) {
 	// Proxied requests have been counted at the activator. Subtract
-	// AverageProxiedConcurrentRequests to avoid double counting.
-	c.buckets.Record(*stat.Time, stat.PodName, stat.AverageConcurrentRequests-stat.AverageProxiedConcurrentRequests)
+	// them to avoid double counting.
+	c.concurrencyBuckets.Record(*stat.Time, stat.PodName, stat.AverageConcurrentRequests-stat.AverageProxiedConcurrentRequests)
+	c.opsBuckets.Record(*stat.Time, stat.PodName, stat.RequestCount-stat.ProxiedRequestCount)
 }
 
 // stableAndPanicConcurrency calculates both stable and panic concurrency based on the
 // current stats.
 func (c *collection) stableAndPanicConcurrency(now time.Time) (float64, float64, error) {
-	spec := c.currentMetric().Spec
+	return c.stableAndPanicStats(now, c.concurrencyBuckets)
+}
 
-	c.buckets.RemoveOlderThan(now.Add(-spec.StableWindow))
+// stableAndPanicConcurrency calculates both stable and panic OPS based on the
+// current stats.
+func (c *collection) stableAndPanicOPS(now time.Time) (float64, float64, error) {
+	return c.stableAndPanicStats(now, c.opsBuckets)
+}
+
+// stableAndPanicStats calculates both stable and panic concurrency based on the
+// given stats buckets.
+func (c *collection) stableAndPanicStats(now time.Time, buckets *aggregation.TimedFloat64Buckets) (float64, float64, error) {
+	spec := c.currentMetric().Spec
+	buckets.RemoveOlderThan(now.Add(-spec.StableWindow))
```

Comment:
Just noticed: we need to redo this kinda. This will leak memory hard if we don‘t ask for specific metrics. Let‘s move the cleanup logic into record?

Reply:
Good catch.

Moved the cleanup to record. I use `time.Now` instead of passing a parameter. `record` may delete some almost outdated stats. Say there's a stat with timestamp 0. A call of `StableAndPanicConcurrency` with timestamp 59.9 is somehow delayed and hasn't read the stat. A call of `record` at 60.1 removes the stat before `StableAndPanicConcurrency` reads. If you think this is not acceptable, I can extend the window size for removing outdated stats.

Reply:
hmm, I think that's acceptable but it makes us dependent on the clock again. WDYT about taking the time of the incoming stat as "now"? Worst case is we delete not enough of the metrics. Either way is kinda harmless I think.

Reply:
Using the same time throughout is probably better.
But this is going to be fun in the window=0 case...

Reply:
Using `stat.Time` makes sense to me. Done.